File: 0112-path-sum/0112-path-sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    
    def hasPathSum(self, root, targetSum):
        self.res = 0
        
        return self.sumItUp(root, targetSum)

    def sumItUp(self, root, targetSum):
        if root == None: 
            return False
        self.res += root.val
        
        if root.right == None and root.left == None:
            if self.res == targetSum:
                return True
            
        if self.sumItUp(root.left, targetSum):
            return True
        if self.sumItUp(root.right, targetSum):
            return True
        
        self.res -= root.val
        return False
        
       
        """
        :type root: TreeNode
        :type targetSum: int
        :rtype: bool
        """
        

File: 0112-path-sum/0112-path-sum.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int num;
int *p = &num;

bool traverse(struct TreeNode* root, int t);
bool hasPathSum(struct TreeNode* root, int targetSum) {
    *p = 0;
    return traverse(root, targetSum);
}

bool traverse(struct TreeNode* root, int t){
    if(root == NULL){
        return false;
    }

    *p += root->val;
    if(root->left != NULL){
        bool ret = traverse(root->left, t);
        if(ret) return true;
    }

    if(root->right != NULL){
        bool ret = traverse( root->right, t);
        if(ret) return true;
    }
    if(*p == t && root->right == NULL && root->left == NULL) return true;
    *p-= root->val;
    return false;

}